### ISSUE TITLE
Explicitly supply pull request ID in user/feed response

### ISSUE THREAD

**Issue by Tathanen** [drone/drone#1784]
Right now if a build is triggered with a `pull_request` status, the pull request ID can be found by parsing the `ref` parameter in `user/feed` and pulling the digits out of the string.  If it's a `push` status that's a PR merge, the ID can be found as part of the `message` parameter.  Searching these strings for the ID can be unreliable, as the user may manipulate or omit these values in some unpredictable way, or they may change with future GitHub or Drone updates.

It'd be great if the pull request ID could be supplied as its own explicit parameter, something like `pr_id`, for more reliable access.  For both the `pull_request` and `push` build types (perhaps set to `null` if it's a push that's not a PR merge).

Related issue that'd benefit from this change: https://github.com/drone/drone-wall/issues/45

**Comment by bradrydzewski** [comment]
the pull request number is not guaranteed to be included in the commit message when a pull request is merged and a push hook is sent by github. The message field is editable in github and in general cannot be relied on:

<img width="412" alt="screen shot 2016-09-02 at 11 50 46 pm" src="https://cloud.githubusercontent.com/assets/817538/18222804/4796fa10-7168-11e6-9083-0fc41f463df4.png">

**Comment by Tathanen** [comment]
Right, which is why I'd rather not rely on it.  Is it supplied by GitHub in any other way?

**Comment by bradrydzewski** [comment]
below is a sample hook received by drone after the pull request is merged. I don't see any structured field container the pull request number

```json
{
  "ref": "refs/heads/master",
  "before": "d4db122c3defcfa32e19f7323aeb1a913a4f182e",
  "after": "23234fa000e1b16853f61e4de4599d1da9a214fb",
  "created": false,
  "deleted": false,
  "forced": false,
  "base_ref": null,
  "compare": "https://github.com/drone/drone/compare/d4db122c3def...23234fa000e1",
  "commits": [
    {
      "id": "a2829a2e963649d32ce2069f7167f649c8f3e39c",
      "tree_id": "1ae64a246994c8b45bb07583be2e4526d18b5263",
      "distinct": true,
      "message": "Add sample gogs web hooks (unused yet)",
      "timestamp": "2016-08-31T22:57:52+02:00",
      "url": "https://github.com/drone/drone/commit/a2829a2e963649d32ce2069f7167f649c8f3e39c",
      "author": {
        "name": "Sandro Santilli",
        "username": "strk"
      },
      "committer": {
        "name": "Sandro Santilli",
        "username": "strk"
      },
      "added": [
        "remote/gogs/testdata/hooks.go"
      ],
      "removed": [

      ],
      "modified": [

      ]
    },
    {
      "id": "bacb78a9e22b1a61cb0a69f0b44ec8a6ed81df1c",
      "tree_id": "5f6fbdb7a54d45cb44897509a4e0e93e43316ddf",
      "distinct": true,
      "message": "Add more hooks to existing fixture file",
      "timestamp": "2016-09-01T17:59:07+02:00",
      "url": "https://github.com/drone/drone/commit/bacb78a9e22b1a61cb0a69f0b44ec8a6ed81df1c",
      "author": {
        "name": "Sandro Santilli",
        "username": "strk"
      },
      "committer": {
        "name": "Sandro Santilli",
        "username": "strk"
      },
      "added": [

      ],
      "removed": [
        "remote/gogs/testdata/hooks.go"
      ],
      "modified": [
        "remote/gogs/fixtures/hooks.go"
      ]
    },
    {
      "id": "320b8b68f0b5013bd403ea71e6d2c2d66aeb421c",
      "tree_id": "23558119002d4cd90096acccf31e7d02fad4fb7b",
      "distinct": true,
      "message": "Fix ticks",
      "timestamp": "2016-09-01T21:49:44+02:00",
      "url": "https://github.com/drone/drone/commit/320b8b68f0b5013bd403ea71e6d2c2d66aeb421c",
      "author": {
        "name": "Sandro Santilli",
        "username": "strk"
      },
      "committer": {
        "name": "Sandro Santilli",
        "username": "strk"
      },
      "added": [

      ],
      "removed": [

      ],
      "modified": [
        "remote/gogs/fixtures/hooks.go"
      ]
    },
    {
      "id": "23234fa000e1b16853f61e4de4599d1da9a214fb",
      "tree_id": "a83354d3371781041bf065f076ff499f3ec4e850",
      "distinct": true,
      "message": "Merge pull request #1778 from strk/gogs-pr-hook\n\nAdd sample gogs web hooks (unused yet) [CI SKIP]",
      "timestamp": "2016-09-01T15:51:56-05:00",
      "url": "https://github.com/drone/drone/commit/23234fa000e1b16853f61e4de4599d1da9a214fb",
      "author": {
        "name": "Brad Rydzewski",
        "username": "bradrydzewski"
      },
      "committer": {
        "name": "GitHub",
        "email": "noreply@github.com",
        "username": "web-flow"
      },
      "added": [

      ],
      "removed": [

      ],
      "modified": [
        "remote/gogs/fixtures/hooks.go"
      ]
    }
  ],
  "head_commit": {
    "id": "23234fa000e1b16853f61e4de4599d1da9a214fb",
    "tree_id": "a83354d3371781041bf065f076ff499f3ec4e850",
    "distinct": true,
    "message": "Merge pull request #1778 from strk/gogs-pr-hook\n\nAdd sample gogs web hooks (unused yet) [CI SKIP]",
    "timestamp": "2016-09-01T15:51:56-05:00",
    "url": "https://github.com/drone/drone/commit/23234fa000e1b16853f61e4de4599d1da9a214fb",
    "author": {
      "name": "Brad Rydzewski",
      "username": "bradrydzewski"
    },
    "committer": {
      "name": "GitHub",
      "email": "noreply@github.com",
      "username": "web-flow"
    },
    "added": [

    ],
    "removed": [

    ],
    "modified": [
      "remote/gogs/fixtures/hooks.go"
    ]
  },
  "repository": {
    "id": 16607898,
    "name": "drone",
    "full_name": "drone/drone",
    "owner": {
      "name": "drone",
      "email": ""
    },
    "private": false,
    "html_url": "https://github.com/drone/drone",
    "description": "Drone is a Continuous Delivery platform built on Docker, written in Go",
    "fork": false,
    "url": "https://github.com/drone/drone",
    "forks_url": "https://api.github.com/repos/drone/drone/forks",
    "keys_url": "https://api.github.com/repos/drone/drone/keys{/key_id}",
    "collaborators_url": "https://api.github.com/repos/drone/drone/collaborators{/collaborator}",
    "teams_url": "https://api.github.com/repos/drone/drone/teams",
    "hooks_url": "https://api.github.com/repos/drone/drone/hooks",
    "issue_events_url": "https://api.github.com/repos/drone/drone/issues/events{/number}",
    "events_url": "https://api.github.com/repos/drone/drone/events",
    "assignees_url": "https://api.github.com/repos/drone/drone/assignees{/user}",
    "branches_url": "https://api.github.com/repos/drone/drone/branches{/branch}",
    "tags_url": "https://api.github.com/repos/drone/drone/tags",
    "blobs_url": "https://api.github.com/repos/drone/drone/git/blobs{/sha}",
    "git_tags_url": "https://api.github.com/repos/drone/drone/git/tags{/sha}",
    "git_refs_url": "https://api.github.com/repos/drone/drone/git/refs{/sha}",
    "trees_url": "https://api.github.com/repos/drone/drone/git/trees{/sha}",
    "statuses_url": "https://api.github.com/repos/drone/drone/statuses/{sha}",
    "languages_url": "https://api.github.com/repos/drone/drone/languages",
    "stargazers_url": "https://api.github.com/repos/drone/drone/stargazers",
    "contributors_url": "https://api.github.com/repos/drone/drone/contributors",
    "subscribers_url": "https://api.github.com/repos/drone/drone/subscribers",
    "subscription_url": "https://api.github.com/repos/drone/drone/subscription",
    "commits_url": "https://api.github.com/repos/drone/drone/commits{/sha}",
    "git_commits_url": "https://api.github.com/repos/drone/drone/git/commits{/sha}",
    "comments_url": "https://api.github.com/repos/drone/drone/comments{/number}",
    "issue_comment_url": "https://api.github.com/repos/drone/drone/issues/comments{/number}",
    "contents_url": "https://api.github.com/repos/drone/drone/contents/{+path}",
    "compare_url": "https://api.github.com/repos/drone/drone/compare/{base}...{head}",
    "merges_url": "https://api.github.com/repos/drone/drone/merges",
    "archive_url": "https://api.github.com/repos/drone/drone/{archive_format}{/ref}",
    "downloads_url": "https://api.github.com/repos/drone/drone/downloads",
    "issues_url": "https://api.github.com/repos/drone/drone/issues{/number}",
    "pulls_url": "https://api.github.com/repos/drone/drone/pulls{/number}",
    "milestones_url": "https://api.github.com/repos/drone/drone/milestones{/number}",
    "notifications_url": "https://api.github.com/repos/drone/drone/notifications{?since,all,participating}",
    "labels_url": "https://api.github.com/repos/drone/drone/labels{/name}",
    "releases_url": "https://api.github.com/repos/drone/drone/releases{/id}",
    "deployments_url": "https://api.github.com/repos/drone/drone/deployments",
    "created_at": 1391759684,
    "updated_at": "2016-09-01T18:29:24Z",
    "pushed_at": 1472763117,
    "git_url": "git://github.com/drone/drone.git",
    "ssh_url": "git@github.com:drone/drone.git",
    "clone_url": "https://github.com/drone/drone.git",
    "svn_url": "https://github.com/drone/drone",
    "homepage": "",
    "size": 33997,
    "stargazers_count": 7504,
    "watchers_count": 7504,
    "language": "Go",
    "has_issues": true,
    "has_downloads": true,
    "has_wiki": true,
    "has_pages": false,
    "forks_count": 859,
    "mirror_url": null,
    "open_issues_count": 109,
    "forks": 859,
    "open_issues": 109,
    "watchers": 7504,
    "default_branch": "master",
    "stargazers": 7504,
    "master_branch": "master",
    "organization": "drone"
  },
  "pusher": {
    "name": "bradrydzewski"
  },
  "organization": {
    "login": "drone",
    "id": 2181346,
    "url": "https://api.github.com/orgs/drone",
    "repos_url": "https://api.github.com/orgs/drone/repos",
    "events_url": "https://api.github.com/orgs/drone/events",
    "hooks_url": "https://api.github.com/orgs/drone/hooks",
    "issues_url": "https://api.github.com/orgs/drone/issues",
    "members_url": "https://api.github.com/orgs/drone/members{/member}",
    "public_members_url": "https://api.github.com/orgs/drone/public_members{/member}",
    "avatar_url": "https://avatars.githubusercontent.com/u/2181346?v=3",
    "description": ""
  },
  "sender": {
    "login": "bradrydzewski",
    "id": 817538,
    "avatar_url": "https://avatars.githubusercontent.com/u/817538?v=3",
    "gravatar_id": "",
    "url": "https://api.github.com/users/bradrydzewski",
    "html_url": "https://github.com/bradrydzewski",
    "followers_url": "https://api.github.com/users/bradrydzewski/followers",
    "following_url": "https://api.github.com/users/bradrydzewski/following{/other_user}",
    "gists_url": "https://api.github.com/users/bradrydzewski/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/bradrydzewski/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/bradrydzewski/subscriptions",
    "organizations_url": "https://api.github.com/users/bradrydzewski/orgs",
    "repos_url": "https://api.github.com/users/bradrydzewski/repos",
    "events_url": "https://api.github.com/users/bradrydzewski/events{/privacy}",
    "received_events_url": "https://api.github.com/users/bradrydzewski/received_events",
    "type": "User",
    "site_admin": false
  }
}
```

**Comment by Tathanen** [comment]
Well dang.  Seems like that should be in there.

So presumably there can only be one open pull request from one user's branch at a time, where all future commits are just added to it.  We'd probably have to hit the GitHub API immediately on/after a commit or merge and check for the most recent PR from that user/branch, and assume that one has the ID we're looking for.

**Comment by Tathanen** [comment]
@bradrydzewski Drone is currently attached to the "pull request" github webook, yes?  In addition to push, which is I believe what you pasted above.  Looking here:

https://developer.github.com/v3/activity/events/types/#pullrequestevent

The pull request event includes the PR ID, and should be triggered every time a PR is "assigned, unassigned, labeled, unlabeled, opened, edited, closed, reopened, or synchronized."  This should get it for most cases, namely when the PR is opened and when a new commit is made to it.  The doc also says:

"If the action is "closed" and the merged key is false, the pull request was closed with unmerged commits. If the action is "closed" and the merged key is true, the pull request was merged."

This actually relates somewhat to another issue I opened recently (https://github.com/drone/drone/issues/1748).  If drone attached itself to the `closed` event, it would know when a merge happened by examining that `merged` key, and be able to report the PR ID via that.  Though really, `closed` would be the fundamentally useful status, so I could remove PRs from the wall even if they were closed without merging.  I talk about that more in that other issue, and what kind of "build" that might be.

So in summary, potential solution involves looking at the pull request hook data to get the ID for inclusion in Drone's response, and also broadcasting during `closed` (and `reopened` ideally) events.

Thanks @projectweekend for the investigation.

**Comment by bradrydzewski** [comment]
I am closing this as out of scope. While I understand how this would be beneficial to the dashboard, it would introduce a level of complexity to the application that is not desirable at this time. Thanks though for the feedback.
